core/arch/arm/pta/sdp_pa.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <mm/core_memprot.h>
+#include <string.h>
+
+#include <sdp_pa.h>
+
+#define PTA_NAME "sdp_pa.pta"
+
+/*
+ * Specficed client TA UUID which has the access to this pta
+ */
+#define TA_SPECFICED_UUID { 0x0b7a14e0, 0xb667, 0x4b3d, \
+		{ 0x84, 0x04, 0xc3, 0xd0, 0xf8, 0xdc, 0x12, 0x44 } }
+
+/*
+ * Specificed TA needs to create pagetable item for data
+ * put in SDP memory. Create a PTA for specficed TA to retrieve
+ * SDP memory physical address.
+ */
+static TEE_Result sdp_pa_cmd_virt_to_phys(uint32_t types,
+			     TEE_Param params[TEE_NUM_PARAMS])
+{
+	char *src = params[0].memref.buffer;
+	size_t sz = params[0].memref.size;
+	paddr_t phys;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: %x", (unsigned)types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (!core_vbuf_is(CORE_MEM_SDP_MEM, src, sz)) {
+		EMSG("bad memref secure attribute");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	phys = virt_to_phys(src);
+
+	params[1].value.a = (uint32_t)phys;
+	params[1].value.b = (uint32_t)(phys >> 32);
+
+	return TEE_SUCCESS;
+}
+
+/*
+ * Trusted Application Entry Points
+ */
+
+static TEE_Result create_ta(void)
+{
+	DMSG("create entry point for pseudo TA \"%s\"", PTA_NAME);
+	return TEE_SUCCESS;
+}
+
+static void destroy_ta(void)
+{
+	DMSG("destroy entry point for pseudo ta \"%s\"", PTA_NAME);
+}
+
+static TEE_Result open_session(uint32_t nParamTypes __unused,
+		TEE_Param pParams[TEE_NUM_PARAMS] __unused,
+		void **ppSessionContext __unused)
+{
+	struct tee_ta_session *calling_sess = tee_ta_get_calling_session();
+	TEE_UUID specficed_ta_uuid = TA_SPECFICED_UUID;
+
+	/* only specficed TA could access */
+	if (calling_sess &&
+	    (memcmp(&calling_sess->ctx->uuid,
```

Comment:
This is a bit inflexible, wouldn't it make more sense to allow any TA with the `TA_FLAG_SECURE_DATA_PATH` flag set instead?

Reply:
Hi @jenswi-linaro,

This method is come from what had been discussed in former issue #1810 , you can see the detail.

Reply:
I can't find anything about restricting to a particular UUID in #1810.

Reply:
Maybe we have some over interpretation on comment: “Check if your TA really needs the physical address or if an optee service (i.e a pTA) should do the conversion and call some drivers that uses the phys addr for some configuration.”
So we check if a TA is allowed to retrieve PA by its UUID. Do you think UUID checking could be removed here?  Actually, we had thought that TA should not get physical address freely.

And we will add the "memref" restrict checking in the next patch.

Reply:
I agree with Jens, allowing only one UUID is too restrictive. Checking a flag is much better. If you think `TA_FLAG_SECURE_DATA_PATH` doesn't fit (why?), then we should consider adding a new flag.

Reply:
OK, we will remove the restrict UUID, and check TA's TA_FLAG_SECURE_DATA_PATH flag.